File: ex_lista/src_python/PrimoAnterior.py
'''
Ana Cristina Silva de Oliveira - 11965630
BCC - A ex05 Primo anterior
'''

from math import isqrt

def crivo_erastostenes(n: int) -> list[int]:
    if n <= 2:
        print("O menor primo anterior ao valor 2 eh o proprio 2")
        return []
    # inicializando os valores como primos
    eh_primo = [True] * n
    eh_primo[0] = False
    eh_primo[1] = False
    # excluindo os multiplos de cada valor
    for primo in range(2,isqrt(n)+1):
        if eh_primo[primo]:
            for multiplo in range(primo*primo,n,primo):
                eh_primo[multiplo] = False
    
    for i in range(len(eh_primo)-1, -1, -1):
        if eh_primo[i] == True and i < n:
            print("O primo anterior ao numero {} eh o {}".format(n,i))
            break
    


    return [primo for primo in range(n) if eh_primo[primo]]

def main():
    try:
        numero = int(input("Digite um valor inteiro\n"))
    except:
        print("Voce nao digitou um valor numerico. Tente novamente")
        exit()

    crivo_erastostenes(numero)
    
        
if __name__ == "__main__":
    main()

File: ex_lista/src_java/PrimoAnterior.java
/* Ana Cristina Silva de Oliveira - NUSP 11965630
   Turma: BCC - A 
   Exercício 05 Numero primo anterior
*/

public class PrimoAnterior {
    public static void main(String[] args) {
        int numero = 0;
        try {
            System.out.println("Digite um valor inteiro: ");
            numero = EntradaTeclado.leInt();
            if (numero == 2) {
                System.out.printf("O menor primo anterior ao valor 2 eh o proprio 2");
                System.exit(1);
            }

        } catch (Exception e) {
            System.out.println("Voce nao digitou um valor numerico. Tente novamente...");
            System.exit(1);
        }
        boolean[] array_primos = crivoErastostenes(numero);
        display_primos(array_primos, numero);
    }
    // funcao para impressao da mensagem
    private static void display_primos(boolean[] array_primos, int numero) {
        for (int i = array_primos.length-1; i >= 2; i--) {
            if (array_primos[i] == true && i < numero) {
                System.out.printf("O menor primo anterior ao valor %d eh o %d", numero, i);
                break; 
            }
        }
    }  
    // funcao que constroi o crivo de erastostenes
    private static boolean[] crivoErastostenes(int numero) {
        boolean[] array_primos = new boolean[numero+1];

        /*setando todos os valores primos de 2 ao numero escolhido como verdadeiros*/

        for (int primo = 2; primo <= numero; primo++) {
            array_primos[primo] = true;
        }
        
        /*excluindo os multiplos de cada valor primo, assim, criando um array 
        apenas com valores primos */
        for (int primo = 2; primo <= numero; primo++) {
            if (array_primos[primo]) {
                for (int multiplo = 2; primo*multiplo <= numero; multiplo++) {
                    array_primos[primo*multiplo] = false;
                }
            }
        }
        
        return array_primos;
    }
}
